File: android/sdk/src/main/java/com/viant/forgeandroid/runtime/ReportBuilderPredicates.kt
package com.viant.forgeandroid.runtime

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull

fun lowerReportBuilderPredicates(config: DashboardReportBuilderDef): DashboardReportBuilderDef {
    if (config.predicates.isEmpty()) return config

    val normalizedPredicates = config.predicates.mapNotNull(::normalizePredicate)
    if (normalizedPredicates.isEmpty()) return config

    val staticFilters = config.staticFilters.toMutableList()
    val staticIds = staticFilters.mapNotNull { (it.id ?: "").trim().takeIf(String::isNotEmpty) }.toMutableSet()
    normalizedPredicates.filter { it.pinned }.forEach { predicate ->
        if (staticIds.add(predicate.id)) {
            staticFilters += lowerPinnedPredicate(predicate)
        }
    }

    val dynamicGroups = config.dynamicFilterGroups.map {
        it.copy(filters = it.filters.toList())
    }.toMutableList()
    val groupById = dynamicGroups.mapNotNull { group ->
        group.id?.trim()?.takeIf { it.isNotEmpty() }?.let { it to group }
    }.toMap().toMutableMap()
    val bucketById = config.predicateBuckets.mapIndexedNotNull { index, bucket ->
        val id = bucket.id?.trim()?.takeIf { it.isNotEmpty() } ?: return@mapIndexedNotNull null
        id to bucket.copy(order = bucket.order ?: index.toDouble())
    }.toMap()

    fun seedGroup(groupId: String): ReportBuilderDynamicFilterGroupDef? {
        val id = groupId.trim()
        if (id.isEmpty()) return null
        groupById[id]?.let { return it }
        val bucket = bucketById[id]
        val seeded = if (bucket != null) {
            ReportBuilderDynamicFilterGroupDef(
                id = id,
                label = bucket.label?.trim()?.ifEmpty { null } ?: pascalCase(id),
                description = bucket.description?.trim()?.ifEmpty { null },
                filters = emptyList()
            )
        } else {
            ReportBuilderDynamicFilterGroupDef(
                id = id,
                label = when (id) {
                    "include" -> "Include"
                    "exclude" -> "Exclude"
                    else -> pascalCase(id)
                },
                filters = emptyList()
            )
        }
        dynamicGroups += seeded
        groupById[id] = seeded
        return seeded
    }

    bucketById.values.sortedBy { it.order ?: 0.0 }.forEach { bucket ->
        bucket.id?.let(::seedGroup)
    }

    val familyMembership = linkedMapOf<String, Pair<MutableList<String>, MutableList<String>>>()
    fun registerFamily(predicate: NormalizedPredicate, direction: NormalizedDirection) {
        if (predicate.group.isBlank() || direction.direction.isBlank()) return
        val filterId = direction.filter.id ?: return
        val membership = familyMembership.getOrPut(predicate.group) { mutableListOf<String>() to mutableListOf() }
        val target = if (direction.direction == "include") membership.first else membership.second
        if (!target.contains(filterId)) target += filterId
    }

    normalizedPredicates.filterNot { it.pinned }.forEach { predicate ->
        predicate.directions.forEach directionLoop@ { direction ->
            val group = seedGroup(direction.groupId) ?: return@directionLoop
            if (group.filters.none { (it.id ?: "").trim() == direction.filter.id }) {
                val updated = group.copy(filters = group.filters + direction.filter)
                val index = dynamicGroups.indexOfFirst { (it.id ?: "").trim() == (group.id ?: "").trim() }
                if (index >= 0) dynamicGroups[index] = updated
                groupById[(updated.id ?: "").trim()] = updated
            }
            registerFamily(predicate, direction)
        }
    }

    val dynamicFamilies = config.dynamicFilterFamilies.toMutableList()
    val existingFamilyIds = dynamicFamilies.mapNotNull { it.id?.trim()?.takeIf(String::isNotEmpty) }.toMutableSet()
    val declaredGroups = config.predicateGroups.mapIndexedNotNull { index, group ->
        val id = group.id?.trim()?.takeIf { it.isNotEmpty() } ?: return@mapIndexedNotNull null
        id to group.copy(order = group.order ?: index.toDouble())
    }
    val declaredById = declaredGroups.toMap()
    val orderedFamilyIds = declaredGroups.map { it.first } +
        familyMembership.keys.filter { id -> declaredById[id] == null }
    orderedFamilyIds.forEach { groupId ->
        val membership = familyMembership[groupId] ?: return@forEach
        if (!existingFamilyIds.add(groupId)) return@forEach
        val declared = declaredById[groupId]
        dynamicFamilies += ReportBuilderDynamicFilterFamilyDef(
            id = groupId,
            label = declared?.label?.trim()?.ifEmpty { null } ?: pascalCase(groupId),
            icon = declared?.icon?.trim()?.ifEmpty { null },
            description = declared?.description?.trim()?.ifEmpty { null },
            includeFilterIds = membership.first,
            excludeFilterIds = membership.second
        )
    }

    return config.copy(
        staticFilters = staticFilters,
        dynamicFilterGroups = dynamicGroups,
        dynamicFilterFamilies = dynamicFamilies
    )
}

private data class NormalizedPredicate(
    val id: String,
    val label: String,
    val description: String,
    val group: String,
    val kind: String,
    val pinned: Boolean,
    val required: Boolean,
    val multiple: Boolean?,
    val presentation: String,
    val semanticRef: String,
    val options: List<ReportBuilderStaticFilterOptionDef>,
    val defaultValue: JsonElement?,
    val paramPath: String,
    val startParamPath: String,
    val endParamPath: String,
    val bucket: String,
    val directions: List<NormalizedDirection>
)

private data class NormalizedDirection(
    val direction: String,
    val groupId: String,
    val filter: ReportBuilderDynamicFilterDef
)

private fun normalizePredicate(entry: ReportBuilderPredicateDef): NormalizedPredicate? {
    val id = entry.id?.trim()?.takeIf { it.isNotEmpty() } ?: return null
    val kind = if ((entry.kind ?: "").trim().equals("dateRange", ignoreCase = true)) "dateRange" else "value"
    val pinned = entry.pinned == true || kind == "dateRange"
    val predicate = NormalizedPredicate(
        id = id,
        label = entry.label?.trim()?.ifEmpty { null } ?: id,
        description = entry.description?.trim().orEmpty(),
        group = entry.group?.trim().orEmpty(),
        kind = kind,
        pinned = pinned,
        required = entry.required == true,
        multiple = entry.multiple,
        presentation = entry.presentation?.trim().orEmpty(),
        semanticRef = entry.semanticRef?.trim().orEmpty(),
        options = entry.options,
        defaultValue = entry.defaultValue,
        paramPath = entry.paramPath?.trim().orEmpty(),
        startParamPath = entry.startParamPath?.trim().orEmpty(),
        endParamPath = entry.endParamPath?.trim().orEmpty(),
        bucket = entry.bucket?.trim()?.ifEmpty { null } ?: "scope",
        directions = emptyList()
    )
    if (pinned) return predicate

    val directions = mutableListOf<NormalizedDirection>()
    val hasInclude = directionEnabled(entry.include)
    val hasExclude = directionEnabled(entry.exclude)
    if (hasInclude) directions += buildDirection(entry, predicate, "include", entry.include)
    if (hasExclude) directions += buildDirection(entry, predicate, "exclude", entry.exclude)
    if (!hasInclude && !hasExclude) directions += buildDirection(entry, predicate, "", null)
    return predicate.copy(directions = directions)
}

private fun lowerPinnedPredicate(predicate: NormalizedPredicate): ReportBuilderStaticFilterDef {
    return ReportBuilderStaticFilterDef(
        id = predicate.id,
        label = predicate.label,
        type = if (predicate.kind == "dateRange") "dateRange" else null,
        required = predicate.required.takeIf { it },
        multiple = predicate.multiple,
        paramPath = predicate.paramPath.takeIf { it.isNotBlank() },
        startParamPath = predicate.startParamPath.takeIf { it.isNotBlank() },
        endParamPath = predicate.endParamPath.takeIf { it.isNotBlank() },
        options = predicate.options,
        defaultValue = predicate.defaultValue
    )
}

private fun buildDirection(
    entry: ReportBuilderPredicateDef,
    predicate: NormalizedPredicate,
    direction: String,
    spec: JsonElement?
): NormalizedDirection {
    val overrides = spec as? JsonObject
    val fallbackId = if (direction.isBlank()) predicate.id else direction + pascalCase(predicate.id)
    val filterId = overrides.string("filterId") ?: fallbackId
    val paramPath = overrides.string("paramPath")
        ?: if (direction.isNotBlank()) {
            when (direction) {
                "include" -> entry.includeParamPath
                "exclude" -> entry.excludeParamPath
                else -> null
            }
        } else {
            entry.paramPath
        }
        ?: "filters.$filterId"
    val multiple = overrides.boolean("multiple") ?: entry.multiple ?: true
    return NormalizedDirection(
        direction = direction,
        groupId = direction.takeIf { it.isNotBlank() } ?: predicate.bucket,
        filter = ReportBuilderDynamicFilterDef(
            id = filterId,
            label = overrides.string("label") ?: predicate.label,
            paramPath = paramPath,
            multiple = multiple,
            emitArray = overrides.boolean("emitArray") ?: entry.emitArray ?: multiple.takeIf { it },
            manualEntry = overrides.boolean("manualEntry") ?: entry.manualEntry,
            manualValueType = overrides.string("manualValueType") ?: entry.manualValueType,
            manualPlaceholder = overrides.string("manualPlaceholder")
                ?: entry.manualPlaceholder
                ?: entry.placeholder,
            dialogId = overrides.string("dialogId") ?: entry.dialogId,
            valueSelector = overrides.string("valueSelector") ?: entry.valueSelector,
            labelSelector = overrides.string("labelSelector") ?: entry.labelSelector,
            groupSelector = overrides.string("groupSelector") ?: entry.groupSelector,
            recordSelectors = overrides.stringList("recordSelectors") ?: entry.recordSelectors,
            requestMapping = overrides.string("requestMapping") ?: entry.requestMapping,
            targetingFeatureKey = overrides.string("targetingFeatureKey") ?: entry.targetingFeatureKey
        )
    )
}

private fun directionEnabled(value: JsonElement?): Boolean {
    if (value == null) return false
    if (value is JsonPrimitive && value.booleanOrNull == false) return false
    return true
}

private fun JsonObject?.string(key: String): String? =
    (this?.get(key) as? JsonPrimitive)?.contentOrNull?.trim()?.takeIf { it.isNotEmpty() }

private fun JsonObject?.boolean(key: String): Boolean? =
    (this?.get(key) as? JsonPrimitive)?.booleanOrNull

private fun JsonObject?.stringList(key: String): List<String>? {
    val value = this?.get(key) ?: return null
    return when (val any = JsonUtil.elementToAny(value)) {
        is List<*> -> any.mapNotNull { it?.toString()?.trim()?.takeIf(String::isNotEmpty) }
        else -> null
    }
}

private fun pascalCase(value: String): String {
    val normalized = value.trim()
    if (normalized.isEmpty()) return ""
    return normalized.replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }
}

File: ios/Sources/ForgeIOSRuntime/ReportBuilderPredicates.swift
import Foundation

public func lowerReportBuilderPredicates(_ config: DashboardReportBuilderDef) -> DashboardReportBuilderDef {
    guard !config.predicates.isEmpty else { return config }
    let normalizedPredicates = config.predicates.compactMap(normalizePredicate)
    guard !normalizedPredicates.isEmpty else { return config }

    var staticFilters = config.staticFilters
    var staticIds = Set(staticFilters.compactMap { $0.id?.trimmedNonEmpty })
    for predicate in normalizedPredicates where predicate.pinned {
        guard !staticIds.contains(predicate.id) else { continue }
        staticIds.insert(predicate.id)
        staticFilters.append(lowerPinnedPredicate(predicate))
    }

    var dynamicGroups = config.dynamicFilterGroups
    var groupById: [String: ReportBuilderDynamicFilterGroupDef] = Dictionary(
        uniqueKeysWithValues: dynamicGroups.compactMap { group in
            guard let id = group.id?.trimmedNonEmpty else { return nil }
            return (id, group)
        }
    )
    let bucketById: [String: ReportBuilderPredicateBucketDef] = Dictionary(
        uniqueKeysWithValues: config.predicateBuckets.enumerated().compactMap { index, bucket in
            guard let id = bucket.id?.trimmedNonEmpty else { return nil }
            return (
                id,
                ReportBuilderPredicateBucketDef(
                    id: bucket.id,
                    label: bucket.label,
                    description: bucket.description,
                    order: bucket.order ?? Double(index)
                )
            )
        }
    )

    @discardableResult
    func seedGroup(_ groupId: String) -> ReportBuilderDynamicFilterGroupDef? {
        let id = groupId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !id.isEmpty else { return nil }
        if let existing = groupById[id] {
            return existing
        }
        let bucket = bucketById[id]
        let seeded = ReportBuilderDynamicFilterGroupDef(
            id: id,
            label: bucket?.label?.trimmedNonEmpty ?? defaultGroupLabel(id),
            icon: nil,
            description: bucket?.description?.trimmedNonEmpty,
            filters: []
        )
        dynamicGroups.append(seeded)
        groupById[id] = seeded
        return seeded
    }

    bucketById.values.sorted { ($0.order ?? 0) < ($1.order ?? 0) }.forEach { bucket in
        if let id = bucket.id {
            seedGroup(id)
        }
    }

    var familyMembership: [String: (include: [String], exclude: [String])] = [:]
    func registerFamily(predicate: NormalizedReportBuilderPredicate, direction: NormalizedReportBuilderDirection) {
        guard !predicate.group.isEmpty, !direction.direction.isEmpty, let filterId = direction.filter.id else {
            return
        }
        var membership = familyMembership[predicate.group] ?? (include: [], exclude: [])
        if direction.direction == "include" {
            if !membership.include.contains(filterId) {
                membership.include.append(filterId)
            }
        } else {
            if !membership.exclude.contains(filterId) {
                membership.exclude.append(filterId)
            }
        }
        familyMembership[predicate.group] = membership
    }

    for predicate in normalizedPredicates where !predicate.pinned {
        for direction in predicate.directions {
            guard let group = seedGroup(direction.groupId), let groupId = group.id?.trimmedNonEmpty else {
                continue
            }
            if !(group.filters.contains { $0.id?.trimmedNonEmpty == direction.filter.id?.trimmedNonEmpty }) {
                let updated = ReportBuilderDynamicFilterGroupDef(
                    id: group.id,
                    label: group.label,
                    icon: group.icon,
                    description: group.description,
                    filters: group.filters + [direction.filter]
                )
                if let index = dynamicGroups.firstIndex(where: { $0.id?.trimmedNonEmpty == groupId }) {
                    dynamicGroups[index] = updated
                }
                groupById[groupId] = updated
            }
            registerFamily(predicate: predicate, direction: direction)
        }
    }

    var dynamicFamilies = config.dynamicFilterFamilies
    var existingFamilyIds = Set(dynamicFamilies.compactMap { $0.id?.trimmedNonEmpty })
    let declaredGroups = config.predicateGroups.enumerated().compactMap { index, group -> (String, ReportBuilderPredicateGroupDef)? in
        guard let id = group.id?.trimmedNonEmpty else { return nil }
        return (
            id,
            ReportBuilderPredicateGroupDef(
                id: group.id,
                label: group.label,
                description: group.description,
                icon: group.icon,
                order: group.order ?? Double(index)
            )
        )
    }
    let declaredById = Dictionary(uniqueKeysWithValues: declaredGroups)
    let orderedFamilyIds = declaredGroups.map(\.0) + familyMembership.keys.filter { declaredById[$0] == nil }
    for groupId in orderedFamilyIds {
        guard let membership = familyMembership[groupId], !existingFamilyIds.contains(groupId) else {
            continue
        }
        existingFamilyIds.insert(groupId)
        let declared = declaredById[groupId]
        dynamicFamilies.append(
            ReportBuilderDynamicFilterFamilyDef(
                id: groupId,
                label: declared?.label?.trimmedNonEmpty ?? pascalCase(groupId),
                icon: declared?.icon?.trimmedNonEmpty,
                description: declared?.description?.trimmedNonEmpty,
                includeFilterIds: membership.include,
                excludeFilterIds: membership.exclude
            )
        )
    }

    return DashboardReportBuilderDef(
        title: config.title,
        subtitle: config.subtitle,
        hooks: config.hooks,
        filterPresentation: config.filterPresentation,
        showFilterCategoryBar: config.showFilterCategoryBar,
        hiddenDynamicGroupIds: config.hiddenDynamicGroupIds,
        notices: config.notices,
        primaryMeasure: config.primaryMeasure,
        measureSections: config.measureSections,
        measures: config.measures,
        computedMeasures: config.computedMeasures,
        dimensions: config.dimensions,
        staticFilters: staticFilters,
        dynamicFilterGroups: dynamicGroups,
        dynamicFilterFamilies: dynamicFamilies,
        predicateBuckets: config.predicateBuckets,
        predicateGroups: config.predicateGroups,
        predicates: config.predicates,
        resultCategories: config.resultCategories,
        groupBy: config.groupBy,
        unifiedFamilyRows: config.unifiedFamilyRows,
        showResultHeader: config.showResultHeader,
        result: config.result
    )
}

private struct NormalizedReportBuilderPredicate {
    let id: String
    let label: String
    let group: String
    let kind: String
    let pinned: Bool
    let required: Bool
    let multiple: Bool?
    let options: [ReportBuilderStaticFilterOptionDef]
    let defaultValue: JSONValue?
    let paramPath: String
    let startParamPath: String
    let endParamPath: String
    let bucket: String
    let directions: [NormalizedReportBuilderDirection]
}

private struct NormalizedReportBuilderDirection {
    let direction: String
    let groupId: String
    let filter: ReportBuilderDynamicFilterDef
}

private func normalizePredicate(_ entry: ReportBuilderPredicateDef) -> NormalizedReportBuilderPredicate? {
    guard let id = entry.id?.trimmedNonEmpty else { return nil }
    let kind = entry.kind?.trimmedNonEmpty?.caseInsensitiveCompare("dateRange") == .orderedSame ? "dateRange" : "value"
    let pinned = entry.pinned == true || kind == "dateRange"
    let base = NormalizedReportBuilderPredicate(
        id: id,
        label: entry.label?.trimmedNonEmpty ?? id,
        group: entry.group?.trimmedNonEmpty ?? "",
        kind: kind,
        pinned: pinned,
        required: entry.required == true,
        multiple: entry.multiple,
        options: entry.options,
        defaultValue: entry.defaultValue,
        paramPath: entry.paramPath?.trimmedNonEmpty ?? "",
        startParamPath: entry.startParamPath?.trimmedNonEmpty ?? "",
        endParamPath: entry.endParamPath?.trimmedNonEmpty ?? "",
        bucket: entry.bucket?.trimmedNonEmpty ?? "scope",
        directions: []
    )
    if pinned {
        return base
    }
    var directions: [NormalizedReportBuilderDirection] = []
    let hasInclude = directionEnabled(entry.include)
    let hasExclude = directionEnabled(entry.exclude)
    if hasInclude {
        directions.append(buildDirection(entry: entry, predicate: base, direction: "include", spec: entry.include))
    }
    if hasExclude {
        directions.append(buildDirection(entry: entry, predicate: base, direction: "exclude", spec: entry.exclude))
    }
    if !hasInclude && !hasExclude {
        directions.append(buildDirection(entry: entry, predicate: base, direction: "", spec: nil))
    }
    return NormalizedReportBuilderPredicate(
        id: base.id,
        label: base.label,
        group: base.group,
        kind: base.kind,
        pinned: base.pinned,
        required: base.required,
        multiple: base.multiple,
        options: base.options,
        defaultValue: base.defaultValue,
        paramPath: base.paramPath,
        startParamPath: base.startParamPath,
        endParamPath: base.endParamPath,
        bucket: base.bucket,
        directions: directions
    )
}

private func lowerPinnedPredicate(_ predicate: NormalizedReportBuilderPredicate) -> ReportBuilderStaticFilterDef {
    ReportBuilderStaticFilterDef(
        id: predicate.id,
        label: predicate.label,
        type: predicate.kind == "dateRange" ? "dateRange" : nil,
        required: predicate.required ? true : nil,
        multiple: predicate.multiple,
        paramPath: predicate.paramPath.isEmpty ? nil : predicate.paramPath,
        startParamPath: predicate.startParamPath.isEmpty ? nil : predicate.startParamPath,
        endParamPath: predicate.endParamPath.isEmpty ? nil : predicate.endParamPath,
        options: predicate.options,
        defaultValue: predicate.defaultValue
    )
}

private func buildDirection(
    entry: ReportBuilderPredicateDef,
    predicate: NormalizedReportBuilderPredicate,
    direction: String,
    spec: JSONValue?
) -> NormalizedReportBuilderDirection {
    let overrides = spec?.objectValue
    let fallbackId = direction.isEmpty ? predicate.id : direction + pascalCase(predicate.id)
    let filterId = overrides.string("filterId") ?? fallbackId
    let explicitParamPath: String?
    if direction == "include" {
        explicitParamPath = entry.includeParamPath
    } else if direction == "exclude" {
        explicitParamPath = entry.excludeParamPath
    } else {
        explicitParamPath = entry.paramPath
    }
    let paramPath = overrides.string("paramPath") ?? explicitParamPath?.trimmedNonEmpty ?? "filters.\(filterId)"
    let multiple = overrides.bool("multiple") ?? entry.multiple ?? true
    return NormalizedReportBuilderDirection(
        direction: direction,
        groupId: direction.isEmpty ? predicate.bucket : direction,
        filter: ReportBuilderDynamicFilterDef(
            id: filterId,
            label: overrides.string("label") ?? predicate.label,
            paramPath: paramPath,
            multiple: multiple,
            emitArray: overrides.bool("emitArray") ?? entry.emitArray ?? (multiple ? true : nil),
            manualEntry: overrides.bool("manualEntry") ?? entry.manualEntry,
            manualValueType: overrides.string("manualValueType") ?? entry.manualValueType,
            manualPlaceholder: overrides.string("manualPlaceholder") ?? entry.manualPlaceholder ?? entry.placeholder,
            dialogId: overrides.string("dialogId") ?? entry.dialogId,
            valueSelector: overrides.string("valueSelector") ?? entry.valueSelector,
            labelSelector: overrides.string("labelSelector") ?? entry.labelSelector,
            groupSelector: overrides.string("groupSelector") ?? entry.groupSelector,
            recordSelectors: overrides.stringList("recordSelectors") ?? entry.recordSelectors,
            requestMapping: overrides.string("requestMapping") ?? entry.requestMapping,
            targetingFeatureKey: overrides.string("targetingFeatureKey") ?? entry.targetingFeatureKey
        )
    )
}

private func directionEnabled(_ value: JSONValue?) -> Bool {
    guard let value else { return false }
    if case .bool(false) = value {
        return false
    }
    return true
}

private func defaultGroupLabel(_ id: String) -> String {
    switch id {
    case "include":
        return "Include"
    case "exclude":
        return "Exclude"
    default:
        return pascalCase(id)
    }
}

private func pascalCase(_ value: String) -> String {
    let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
    guard let first = trimmed.first else { return "" }
    return String(first).uppercased() + trimmed.dropFirst()
}

private extension Optional where Wrapped == [String: JSONValue] {
    func string(_ key: String) -> String? {
        self?[key]?.stringValue?.trimmedNonEmpty
    }

    func bool(_ key: String) -> Bool? {
        self?[key]?.boolValue
    }

    func stringList(_ key: String) -> [String]? {
        guard let value = self?[key] else { return nil }
        switch value {
        case .array(let values):
            return values.compactMap { $0.stringValue?.trimmedNonEmpty }
        default:
            return nil
        }
    }
}

private extension String {
    var trimmedNonEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}
